src/mlpack/tests/main_tests/mean_shift_test.cpp

```diff
@@ -0,0 +1,176 @@
+/**
+ * @file mean_shift_test.cpp
+ * @author Tan Jun An
+ *
+ * Test mlpackMain() of mean_shift_main.cpp.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#include <string>
+
+#define BINDING_TYPE BINDING_TYPE_TEST
+static const std::string testName = "MeanShift";
+
+#include <mlpack/core.hpp>
+#include <mlpack/core/util/mlpack_main.hpp>
+#include <mlpack/methods/mean_shift/mean_shift_main.cpp>
+#include "test_helper.hpp"
+
+#include <boost/test/unit_test.hpp>
+#include "../test_tools.hpp"
+
+using namespace mlpack;
+
+struct MeanShiftTestFixture
+{
+ public:
+  MeanShiftTestFixture()
+  {
+    // Cache in the options for this program.
+    CLI::RestoreSettings(testName);
+  }
+
+  ~MeanShiftTestFixture()
+  {
+    // Clear the settings.
+    bindings::tests::CleanMemory();
+    CLI::ClearSettings();
+  }
+};
+
+static void ResetSettings()
+{
+  bindings::tests::CleanMemory();
+  CLI::ClearSettings();
+  CLI::RestoreSettings(testName);
+}
+
+BOOST_FIXTURE_TEST_SUITE(MeanShiftMainTest, MeanShiftTestFixture);
+
+/**
+ * Ensure that the output has 1 extra row for the labels and
+ * check the number of points for output remain the same.
+ */
+BOOST_AUTO_TEST_CASE(MeanShiftOutputDimensionTest)
+{
+  arma::mat x;
+  x.randu(3,100); // 100 points in 3 dimension
+
+  // Input random data points.
+  SetInputParam("input", std::move(x));
+
+  mlpackMain();
+
+  // Now check that the output has 1 extra row for labels.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("output").n_rows, 3 + 1);
+  // Check number of output points are the same.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("output").n_cols, 100);
+}
+
+/**
+ * Ensure that if we ask for labels_only, output has 1 column and
+ * same number of rows for each point's label.
+ */
+BOOST_AUTO_TEST_CASE(MeanShiftLabelOnlyOutputDimensionTest)
+{
+  arma::mat x;
+  x.randu(3,100); // 100 points in 3 dimension
+
+  // Input random data points.
+  SetInputParam("input", std::move(x));
+  SetInputParam("labels_only", true);
+
+  mlpackMain();
+
+  // Check number of output points are the same.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("output").n_rows, 100);
+  // Check that there is only 1 column containing all the labels.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("output").n_cols, 1);
```

Comment:
Hm, it seems strange to me that this is transposed (I would expect one column and 100 rows).  If you are willing to dig a little deeper, I see that the issue is that `Cluster()` returns an `arma::Col<size_t>` for assignments, but really this should be an `arma::Row<size_t>`.  It will require a little refactoring, then we will need to change this test a bit and ensure that when we run `mlpack_mean_shift` from the command-line with `--output_file` and `--labels_only` set, we get a file with many rows and only one column.

I also tested `mlpack_mean_shift` from the command-line with `--output_file` and `--labels_only`, and it puts all the labels in a single line of an output CSV, which is something that should be fixed.

Don't feel obligated to fix the issue; I can also open an issue if you prefer.  Just let me know if you'd like to handle it here in this PR or not.

Reply:
@rcurtin 

I tried converting the `arma::Col<size_t>` to `arma::Row<size_t>` in `mean_shift_impl.hpp` and `mean_shift_main.cpp` but it doesn't seem to work.
Rather, I tried just removing the `.t()` from `assignments = resultingNeighbours.t()` at the end of `mean_shift_impl.hpp` and it seems to resolve this issue. 
So should I still refactor each `arma::Col<size_t>` to `arma::Row<size_t>` then?